### ISSUE TITLE
 MajorLeagueCyber login should be togglable

### ISSUE THREAD

**Issue by sachinkamath** [CTFd/CTFd#1063]
Right now, CTFd has MajorLeagueCyber login by default. I'd like to see this as an option which can be enabled in configuration.

If that's the default behavior you want @ColdHeat , maybe there can be a toggle to turn off  MajorLeagueCyber register and login from the platform. 

This change has to be manually be done by editing the register and login pages at the moment.

**Comment by frankli0324** [comment]
```patch
From 833eb089ca002639e7eb1e21e2eade554d92b175 Mon Sep 17 00:00:00 2001
From: Frank <frankli0324@hotmail.com>
Date: Mon Sep 17 00:00:00 2001
Subject: [PATCH] adds a config option to disable MLC

---
 CTFd/auth.py                                 | 11 +++++++----
 CTFd/themes/admin/templates/configs/mlc.html | 15 +++++++++++++--
 CTFd/themes/core/templates/login.html        |  2 ++
 CTFd/themes/core/templates/register.html     |  2 ++
 CTFd/views.py                                |  3 +++
 5 files changed, 27 insertions(+), 6 deletions(-)

diff --git a/CTFd/auth.py b/CTFd/auth.py
index c937031..0a6e5a7 100644
--- a/CTFd/auth.py
+++ b/CTFd/auth.py
@@ -203,6 +203,7 @@ def register():
                 name=request.form["name"],
                 email=request.form["email"],
                 password=request.form["password"],
+                mlc_disabled=get_config("mlc_disabled"),
             )
         else:
             with app.app_context():
@@ -246,7 +247,7 @@ def register():
 
         return redirect(url_for("challenges.listing"))
     else:
-        return render_template("register.html", errors=errors)
+        return render_template("register.html", errors=errors, mlc_disabled=get_config("mlc_disabled"))
 
 
 @auth.route("/login", methods=["POST", "GET"])
@@ -281,20 +282,22 @@ def login():
                 log("logins", "[{date}] {ip} - submitted invalid password for {name}")
                 errors.append("Your username or password is incorrect")
                 db.session.close()
-                return render_template("login.html", errors=errors)
+                return render_template("login.html", errors=errors, mlc_disabled=get_config("mlc_disabled"))
         else:
             # This user just doesn't exist
             log("logins", "[{date}] {ip} - submitted invalid account information")
             errors.append("Your username or password is incorrect")
             db.session.close()
-            return render_template("login.html", errors=errors)
+            return render_template("login.html", errors=errors, mlc_disabled=get_config("mlc_disabled"))
     else:
         db.session.close()
-        return render_template("login.html", errors=errors)
+        return render_template("login.html", errors=errors, mlc_disabled=get_config("mlc_disabled"))
 
 
 @auth.route("/oauth")
 def oauth_login():
+    if get_config("mlc_disabled"):
+        return redirect("/login")
     endpoint = (
         get_app_config("OAUTH_AUTHORIZATION_ENDPOINT")
         or get_config("oauth_authorization_endpoint")
diff --git a/CTFd/themes/admin/templates/configs/mlc.html b/CTFd/themes/admin/templates/configs/mlc.html
index aac41da..b668b7e 100644
--- a/CTFd/themes/admin/templates/configs/mlc.html
+++ b/CTFd/themes/admin/templates/configs/mlc.html
@@ -25,6 +25,17 @@
 				   {% if oauth_client_secret is defined and oauth_client_secret != None %}value="{{ oauth_client_secret }}"{% endif %}>
 		</div>
 
-		<button type="submit" class="btn btn-md btn-primary float-right">Update</button>
-	</form>
+        <div class="form-group form-check">
+            <label>
+                <input id="mlc_disabled" name="mlc_disabled" type="checkbox" {% if mlc_disabled %}checked{% endif %}>
+                Disable MajorLeagueCyber<br>
+                <small class="form-text text-muted">
+                    Disable <a href="https://majorleaguecyber.org/">MajorLeagueCyber</a> integration in this instance
+                </small>
+            </label>
+        </div>
+
+
+        <button type="submit" class="btn btn-md btn-primary float-right">Update</button>
+    </form>
 </div>
\ No newline at end of file
diff --git a/CTFd/themes/core/templates/login.html b/CTFd/themes/core/templates/login.html
index a31b973..a2eb5a3 100644
--- a/CTFd/themes/core/templates/login.html
+++ b/CTFd/themes/core/templates/login.html
@@ -20,10 +20,12 @@
 				</div>
 			{% endfor %}
 
+            {% if not mlc_disabled %}
 			<a class="btn btn-secondary btn-lg btn-block" href="{{ url_for('auth.oauth_login') }}">Login with Major
 				League Cyber</a>
 
 			<hr>
+            {% endif %}
 
 			<form method="post" accept-charset="utf-8" autocomplete="off" role="form" class="form-horizontal">
 				<div class="form-group">
diff --git a/CTFd/themes/core/templates/register.html b/CTFd/themes/core/templates/register.html
index 21dc3f2..3950290 100644
--- a/CTFd/themes/core/templates/register.html
+++ b/CTFd/themes/core/templates/register.html
@@ -20,9 +20,11 @@
 				</div>
 			{% endfor %}
 
+            {% if not mlc_disabled %}
 			<a class="btn btn-secondary btn-lg btn-block" href="{{ url_for('auth.oauth_login') }}">Login with Major League Cyber</a>
 
 			<hr>
+            {% endif %}
 
 			<form method="post" accept-charset="utf-8" autocomplete="off" role="form" class="form-horizontal">
 				<div class="form-group">
diff --git a/CTFd/views.py b/CTFd/views.py
index 2cc7110..91eca0f 100644
--- a/CTFd/views.py
+++ b/CTFd/views.py
@@ -107,6 +107,9 @@ def setup():
             set_config("mail_password", None)
             set_config("mail_useauth", None)
 
+            # MajorLeagueCyber
+            set_config("mlc_disabled", False)
+
             set_config("setup", True)
 
             try:
-- 
2.22.0
```

**Comment by frankli0324** [comment]
applying the patch would add an option in the admin config to disable MLC

**Comment by sachinkamath** [comment]
Hi Frank,

I think this would be better as a PR so that it can just be merged right
away to the codebase :)

**Comment by frankli0324** [comment]
did, https://github.com/CTFd/CTFd/pull/1075  
however, I thought for a few minutes and then decided that MLC may be is more than a feature... so I closed the PR myself.
the pr didn't pass the checks anyway (

**Comment by pintman** [comment]
I agree to make the feature configurable and to allow running ctfd on its own without integration into mlc.

**Comment by jainendra** [comment]
I agree that it should be a feature

**Comment by sir-ragna** [comment]
I opened a pull request with the code from @frankli0324 on the 2.2.0-dev branch and made it pass the checks. 

![disable_mcl2](https://user-images.githubusercontent.com/1280559/67423018-cb58a680-f5d3-11e9-8e2d-7660398d8cb1.gif)

**Comment by Hedroed** [comment]
+1

**Comment by ColdHeat** [comment]
This should now be resolved in 2.2.0-dev which has a different setup flow for MLC.
